pytensor/tensor/rewriting/linalg.py

```diff
@@ -377,3 +381,31 @@ def local_lift_through_linalg(
                 return [block_diag(*inner_matrices)]
             else:
                 raise NotImplementedError  # pragma: no cover
+
+
+@register_canonicalize
+@register_stabilize
+@register_specialize
+@node_rewriter([SVD])
+def local_svd_uv_simplify(fgraph, node):
+    """If we have more than one `SVD` `Op`s and at least one has keyword argument
+    `compute_uv=True`, then we can change `compute_uv = False` to `True` everywhere
+    and allow `pytensor` to re-use the decomposition outputs instead of recomputing.
+    """
+    (x,) = node.inputs
+    svd_count = 0
+    compute_uv = False
+    not_compute_uv_svd_list = []
+
+    for cl, _ in fgraph.clients[x]:
+        if isinstance(cl.op, Blockwise) and isinstance(cl.op.core_op, SVD):
+            svd_count += 1
+            if (not compute_uv) and cl.op.core_op.compute_uv:
+                compute_uv = True
+            if not cl.op.core_op.compute_uv:
+                not_compute_uv_svd_list.append(cl)
+
+    if svd_count > 1 and compute_uv:
+        for cl in not_compute_uv_svd_list:
+            cl.op.core_op.compute_uv = True
+    return [cl.outputs[0] for cl in not_compute_uv_svd_list]
```

Comment:
I think changing properties of the op inplace might lead to problems...

This rewrite function should run for each SVD node, so maybe it is easier to just locate an existing `compute_uv = True` node, and return that as replacement for each `compuet_uv = False` node?

So something like:

- If `compute_uv` is False, return and do nothing
- check if there is a `compute_uv = True` node in the graph with the same input. If not, return and do nothing
- Return the exising output of that node as replacement for the current `compute_uv = False` node.

I wonder though if there could be bad interactions somewhere if there is a rewrite that replaces `compute_uv = Fales` nodes if they are not used? We don't want to run into any infinite cycles...

Reply:
@ricardoV94 Do you know if there are any problems that could happen if a rewrite returns an existing variable instead of a new one?

Reply:
I think there will be a problem only when a rewrite tries to replace a variable by another that depends on the original variable.

Reply:
And yes we shouldn't modify the properties in place. We should replace the smaller Op by the bigger one, just make sure the smaller one is not in the ancestors of the bigger one.

Reply:
Otherwise creating a new SVD should be simple, just call the user facing constructor with the specific flags

Reply:
I realized that I was trying to rewrite the whole graph instead of just the SVD node passed to the rewriter. I changed the code logic and committed.

Reply:
I changed the codes. I was essentially trying to rewrite the whole graph with the code.
I observed something using this small example:
```py
import pytensor
import pytensor.tensor as pt
from pytensor.tensor.type import matrix
from pytensor.tensor.linalg import svd

a_pt = matrix("a")
s = svd(a_pt, full_matrices=False, compute_uv=False)
gs = pt.grad(pt.sum(s), a_pt)
f = pytensor.function([a_pt], gs)
e = pytensor.graph.fg.FunctionGraph([a_pt], [gs], clone=False)
```
The graph for `f` has only 1 SVD Ops with `compute_uv=True`:
![image](https://github.com/pymc-devs/pytensor/assets/97870091/4b12ba16-5d9c-47d7-8c10-a4fa0f31c480)

But the graph for `e` has 2 SVD Ops with different `compute_uv`:
![image](https://github.com/pymc-devs/pytensor/assets/97870091/2e170b9f-6a44-4ed4-a24f-b1e67a4d28e8)

I wondered 2 things:
1. I thought that `e` graph is supposed to be the rewritten one?
2. If `f`'s graph is really the rewritten one, then there may be logic somewhere for the rewrite before I touched it since the 2 graphs still look exactly like above when I comment my added codes out 😓.

Reply:
It will arise in gradient graphs. For example, you can just do:

```
X = pt.dmatrix('X')
s = pt.linalg.svd(X, compute_uv=False)
g = pt.grad(s, X)
```

The graph for `g` will re-compute the SVD of `X` during the backward pass with `compute_uv = True`, because we require the matrices `U` and `V` to compute the gradient of `s` with respect to `X`. Pytensor then won't be able to see that these two computations are the same, and will end up computing the SVD twice.

Reply:
```py
a_pt = matrix("a")
s = svd(a_pt, full_matrices=False, compute_uv=False)
gs = pt.grad(pt.sum(s), a_pt)
f = pytensor.function([a_pt], gs)
e = pytensor.graph.fg.FunctionGraph([a_pt], [gs], clone=False)
```
Thank you. I indeed received a graph for `gs` and `e` with 2 different SVD:
![image](https://github.com/pymc-devs/pytensor/assets/97870091/c3c13c56-7fc1-493a-a37a-d00600123e35)

But for `f`, I receive a graph with just a single SVD:
![image](https://github.com/pymc-devs/pytensor/assets/97870091/9b9adadd-1c96-4aa2-b84c-a50105bc5a3d)

1. Which graph will be used in calculation if I run `f([[1, 2], [3, 4]])`?
2. Which graph will be used in calculation if I run `gs.eval({a_pt : [[1, 2], [3, 4]]})`?

Reply:
BTW, this is `f` summary profile:
```sh
Function profiling
==================
  Message: /tmp/ipykernel_1282122/871230895.py:10
  Time in 1 calls to Function.__call__: 3.448710e-02s
  Time in Function.vm.__call__: 0.03426380921155214s (99.353%)
  Time in thunks: 0.03424406051635742s (99.295%)
  Total compilation time: 4.109558e-02s
    Number of Apply nodes: 2
    PyTensor rewrite time: 2.893809e-02s
       PyTensor validate time: 2.457825e-04s
    PyTensor Linker time (includes C, CUDA code generation/compiling): 0.00876139895990491s
       C-cache preloading 5.506449e-03s
       Import time 8.061258e-04s
       Node make_thunk time 1.967770e-03s
           Node Dot22(SVD{full_matrices=False, compute_uv=True}.0, SVD{full_matrices=False, compute_uv=True}.2) time 1.942240e-03s
           Node SVD{full_matrices=False, compute_uv=True}(a) time 1.436425e-05s

Time in all call to pytensor.grad() 1.036228e-02s
Time since pytensor import 2.774s
Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  99.8%    99.8%       0.034s       3.42e-02s     Py       1       1   pytensor.tensor.nlinalg.SVD
   0.2%   100.0%       0.000s       6.60e-05s     C        1       1   pytensor.tensor.blas.Dot22
   ... (remaining 0 Classes account for   0.00%(0.00s) of the runtime)

Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  99.8%    99.8%       0.034s       3.42e-02s     Py       1        1   SVD{full_matrices=False, compute_uv=True}
   0.2%   100.0%       0.000s       6.60e-05s     C        1        1   Dot22
   ... (remaining 0 Ops account for   0.00%(0.00s) of the runtime)

Apply
------
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
  99.8%    99.8%       0.034s       3.42e-02s      1     0   SVD{full_matrices=False, compute_uv=True}(a)
   0.2%   100.0%       0.000s       6.60e-05s      1     1   Dot22(SVD{full_matrices=False, compute_uv=True}.0, SVD{full_matrices=False, compute_uv=True}.2)
   ... (remaining 0 Apply instances account for 0.00%(0.00s) of the runtime)
```